keps/sig-storage/20190122-csi-inline-volumes.md

```diff
@@ -115,46 +115,61 @@ This proposal introduces the support of both persistent and ephemeral-style volu
+A CSI driver will have to indicate the inline volume lifecycle, that it supports, as either `Persistent` or `Ephemeral`.  This setting will be stored in a [`CSIDriver` configuration CRD](https://github.com/kubernetes/enhancements/issues/594) and will be checked at runtime by Kubelet and external CSI components for enforcement.  The setting is to inform the Kubelet and other CSI components how to process inline volume information for proper setup.
 
-#### Persistent inline mode
-For CSI drivers that indicate they support persistent inline volumes, the `CSIDriver.lifecycle` value should be set to `persistent` (default of if not provided).  When a driver is used to handle volumes from an inline context: 
-* Volumes are assumed to be originated from an inline pod spec.
+### Persistent inline mode
+For CSI drivers that indicate they support persistent inline volume definitions, the `CSIDriver.Lifecycle` value should be set to `Persistent` (default if not provided).  When a persistent inline driver is used to handle volume operations: 
+* Volume requests are assumed to have originated from an inline pod spec.
 * The `volumeHandle` value must be generated ahead of time and manually entered in the pod spec.
-* The driver will only receive CSI calls related to storage opearations for *attach*, *mount*, *unmount*, *detach*.
-* The driver will not receive CSI calls for *porovision* and *deletion* of volumes.
-
-#### Ephemeral inline mode
-Ephemeral inline volumes are more restricted than their persistent counterparts.  The `CSIDriver.lifecycle = ephemeral` must be set to indicate support for inline ephemeral volumes which will follow the lifecycle of its pod.  For drivers indicating support for ephemral mode: 
-* Volumes are assumed to be originated from an inline pod spec.
-* The `volumeHandle` is not required and is ignored if provided in the pod spec.
-* CSI will internally generate a `volumeHandle` value which is passed to the driver.
+* Omission of the volumeHandle will result in an error during the attachment phase which will fail.
+* The driver will only receive CSI calls related to storage operations for *attach*, *mount*, *unmount*, *detach*.
+* The driver will not receive CSI calls for *provision* and *deletion* of volumes.
+
+#### Persistent inline volume limitations
+During the attachment phase, information about pod and pod namespace is not available.  This introduces the following limitations for persistent inline volumes:
+* Persistent inline volume requests will not participate in attach/detach operations that require secrets.
+* Access to pod namespaces are not available during attachment.
```

Comment:
> Access to pod namespaces are not available during attachment.

does this mean we can omit the namespace from VolumeAttachmentSource and flatten/simplify it to this:

```golang
type VolumeAttachmentSource struct {
  // PersistentVolumeName is the name of a persistent volume to attach.
  // Mutually exclusive with inlineVolumeSource.
  // +optional
  PersistentVolumeName *string

  // InlineVolumeSource is the definition of an inline volume in a pod to attach.
  // Mutually exclusive with persistentVolumeName.
  InlineVolumeSource *v1.VolumeSource
}
```

Reply:
At first glance the answer is yes.  I will double check but I believe this can be flattened.

Reply:
this comment is still outstanding